### PR TITLE
Generating CLI documentation

### PR DESCRIPTION
**Pull request by rd4398**
Fixes: https://github.com/python-wheel-build/fromager/issues/217

References: 
https://sphinx-click.readthedocs.io/en/latest/examples/commands/
https://www.sphinx-doc.org/en/master/usage/quickstart.html
https://stackoverflow.com/questions/53668052/sphinx-cannot-find-my-python-files-says-no-module-named

### CODE REVIEW

File: docs/index.rst
Comment (tiran) [inline docs/index.rst:18]: The instruction expects a Python import name to the root click group. For Fromager it's `from fromager.__main__ import main`:

```suggestion
.. click:: fromager.__main__:main
```

**Review by tiran — CHANGES_REQUESTED** [review]
It's a good idea and a good start. I have done the same for InstructLab a while ago,

Next steps:

- move the click doc to a separate page, https://github.com/instructlab/instructlab/blob/main/docs/ilab.rst
- configure the click extension, https://github.com/instructlab/instructlab/blob/main/docs/conf.py
- create a requirements.txt in the docs folder, https://github.com/instructlab/instructlab/blob/main/docs/requirements.txt
- create a tox target to build the docs: https://github.com/instructlab/instructlab/blob/main/tox.ini#L124-L134

File: docs/index.rst
Comment (rd4398) [reply docs/index.rst]: I changed this and now I get the following error:

```
/home/rdevasth/fromager/docs/cli.rst:1: ERROR: Failed to import "main" from "fromager.__main__". The following exception was raised:
Traceback (most recent call last):
  File "/home/rdevasth/fromager/venv/lib/python3.11/site-packages/sphinx_click/ext.py", line 444, in _load_module
    mod = __import__(module_name, globals(), locals(), [attr_name])
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'fromager'
```

**Comment by rd4398** [comment]
> It's a good idea and a good start. I have done the same for InstructLab a while ago,
> 
> Next steps:
> 
> * move the click doc to a separate page, https://github.com/instructlab/instructlab/blob/main/docs/ilab.rst
> * configure the click extension, https://github.com/instructlab/instructlab/blob/main/docs/conf.py
> * create a requirements.txt in the docs folder, https://github.com/instructlab/instructlab/blob/main/docs/requirements.txt
> * create a tox target to build the docs: https://github.com/instructlab/instructlab/blob/main/tox.ini#L124-L134

I did first three steps but still I haven't been able to generate the documentation when I run `make html`
I also get a ModuleNotFound error that I mentioned above. I have pushed my latest attempt. Am I still missing something?

File: docs/index.rst
Comment (dhellmann) [reply docs/index.rst]: Do you have fromager installed? Sphinx is going to try to import the source code to generate those docs.

One convenient way to do that is to use tox instead of make. Tox lets you define a virtualenv with dependencies and then a set of commands to run. [This](https://github.com/sphinx-contrib/spelling/blob/main/tox.ini#L72) is what I use for generating documentation for another library that I maintain, for example.

**Comment by rd4398** [comment]
cli
```
.. click:: fromager.__main__:main
  :prog: bootstrap
  :nested: full
```
Not sure whether `:prog: ` should be `bootstrap` since it looks like above. Any idea what it should be?

**Comment by tiran** [comment]
> Not sure whether `:prog: ` should be `bootstrap` since it looks like above. Any idea what it should be?

`prog` should be the name of the script. In our case it is `fromager`.

**Comment by rd4398** [comment]
Ah, okay! Changing that and adding `readthedocs.yaml` file as well

**Comment by rd4398** [comment]
Let me know if any changes are required!

File: tox.ini
Comment (dhellmann) [inline tox.ini:76]: The `--fail-on-warning` flag makes this PR fail to build for me locally because there is no `docs/_static` directory. You could create that directory with an empty file named `.gitkeep` to resolve that.

**Comment by dhellmann** [comment]
This looks good to me. @tiran had asked for a change, I'll wait for him to re-review before merging.

File: docs/conf.py
Comment (tiran) [inline docs/conf.py:12]: Let's use https://docs.python.org/3/library/importlib.metadata.html to get the version from the package metadata.

File: .gitignore
Comment (tiran) [inline .gitignore:27]: Missing newline at EOF

File: docs/requirements.txt
Comment (tiran) [inline docs/requirements.txt:3]: Missing newline at EOF
You also don't need to include the package as dependency. It's implied.

File: readthedocs.yaml
Comment (tiran) [inline readthedocs.yaml:19]: The file has to be named `.readthedocs.yaml` with a leading dot.

Missing newline at EOF

File: tox.ini
Comment (tiran) [inline tox.ini:69]: Missing newline at EOF

**Review by tiran — CHANGES_REQUESTED** [review]

File: docs/conf.py
Comment (rd4398) [reply docs/conf.py:12]: I tried using it but Sphinx is detecting it as function for some reason and I get the error:
`TypeError: expected string or bytes-like object, got 'function'`

Code:
```
from importlib.metadata import version
project = "Fromager"
copyright = "2024, Doug Hellmann"
author = "Doug Hellmann"
release = version("fromager")

```

Comment (tiran) [reply docs/conf.py:12]: I bet something is looking at `version` attribute of the module

do `from importlib import metadata`

**Comment by rd4398** [comment]
I have made the suggested changes!
Let me know if any more changes are required.
Also, @tiran is there any reason to leave a blank line at EOF for some files you mentioned above? Just curious about why do we do that.

**Comment by rd4398** [comment]
Not sure why one check is getting cancelled only after squishing commits :(

**Comment by dhellmann** [comment]
I'm happy with this version. If we spot any other issues, let's plan to submit follow-ups after we have the doc publishing configured.

**Comment by dhellmann** [comment]
> is there any reason to leave a blank line at EOF for some files you mentioned above? Just curious about why do we do that.

If we all configure our editors the same way it cuts down on spurious reformatting changes when we submit PRs. Having a blank line at the end of a file means if you `cat` it then it comes out with your command prompt on the next line instead of joined up with the last line of the file.

**Review by tiran — APPROVED** [review]

**Pull request merged**
